<a href="https://colab.research.google.com/github/NahataiTep/DPDM65/blob/main/pre-processing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

In [2]:
from google.colab import drive # เชื่อม google drive
drive.mount('/content/drive')

Mounted at /content/drive


In [38]:
project_df = pd.read_csv('/content/drive/MyDrive/DPDM2022/Copy of opendata_project.csv')
project_df.head()

row_number                project_id                             name_en  \
0      7880.0  5ee08d64dda791001299b2a6  Winning Residence Sukhumvit Praksa   
1     16689.0             project_11326                    Gift Condominium   
2       290.0  62739c4089233e001362b001    Sanrak 9 Sukprayoon Nong Krathum   
3     20719.0              project_2010                     Baan Sansabai 1   
4       199.0             project_16838          Baanpruksa Ratchapruek-345   

                              name_th propertytype_id propertytype_name_en  \
0  วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1       Detached House   
1                 กิ๊ฟท์ คอนโดมิเนียม               2                Condo   
2      แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3             Townhome   
3                       บ้านแสนสบาย 1               1       Detached House   
4              บ้านพฤกษา ราชพฤกษ์-345               3             Townhome   

  propertytype_name_th  price_min     developer_id  \
0                 บ้าน  3790000.0  developer_10678   
1                คอนโด  1100000.0   developer_7319   
2             ทาวน์โฮม        0.0  developer_10510   
3                 บ้าน        0.0  developer_10510   
4             ทาวน์โฮม  1930000.0   developer_1026   

              developer_name_en  ... facility_meeting  facility_park  \
0  Winning Asset Plus Co., Ltd.  ...              NaN            NaN   
1      Gift Asset Land Co.,Ltd.  ...              NaN            1.0   
2                           NaN  ...              NaN            NaN   
3                           NaN  ...              NaN            NaN   
4  Preuksa Real Estate Co.,Ltd.  ...              NaN            1.0   

   facility_playground facility_pool facility_security         date_created  \
0                  NaN           1.0               1.0  2020-06-10 14:36:04   
1                  NaN           1.0               1.0  2017-10-30 15:38:48   
2                  NaN           NaN               NaN  2022-05-05 16:43:28   
3                  NaN           NaN               NaN  2016-09-05 15:32:15   
4                  NaN           1.0               1.0  2019-09-05 11:49:37   

           date_finish         date_updated  source  \
0                  NaN  2020-12-04 10:32:58  baania   
1  2013-01-01 07:00:00  2020-12-04 10:49:00  baania   
2                  NaN  2022-05-05 17:30:03  baania   
3  1970-01-02 07:00:00  2020-12-04 10:45:31  baania   
4  1970-01-02 07:00:00  2021-03-17 12:00:04  baania   

                                         url_project  
0  https://baania.com/th/project/วินนิ่ง-เรสซิเดน...  
1  https://baania.com/th/project/กิ๊ฟท์-คอนโดมิเน...  
2  https://baania.com/th/project/แสนรัก-9-ศุขประย...  
3  https://baania.com/th/project/บ้านแสนสบาย-1-pr...  
4  https://baania.com/th/project/บ้านพฤกษา-ราชพฤก...  

[5 rows x 45 columns]

In [4]:
unit_df = pd.read_csv('/content/drive/MyDrive/DPDM2022/Copy of opendata_unittype.csv')
unit_df.head()

row_number     unittype_id    project_id             name_th  \
0       27501   unittype_9565  project_2446           2 Bedroom   
1       28777  unittype_10267  project_3543  บ้านแฝดสองชั้น Z-2   
2       28865    unittype_471   project_143    1 Bedroom Type B   
3       24988  unittype_19506  project_7766    Studio Floor 1-2   
4       38596  unittype_10290  project_3551                TH 3   

   propertytype_id propertytype_name_en propertytype_name_th  \
0                2                Condo                คอนโด   
1            20000           Twin house              บ้านแฝด   
2                2                Condo                คอนโด   
3                2                Condo                คอนโด   
4                3             Townhome             ทาวน์โฮม   

   area_total_min_wa  area_usable_min  price_min  ...  count_room_guest  \
0                NaN             37.3  2500000.0  ...               1.0   
1               40.0            150.0  2390000.0  ...               NaN   
2                NaN             28.0        NaN  ...               NaN   
3                NaN             27.0   170000.0  ...               NaN   
4                NaN            182.0  2390000.0  ...               1.0   

   count_room_kitchen  count_room_living  count_room_maid  count_room_storage  \
0                 1.0                NaN              NaN                 NaN   
1                 NaN                NaN              NaN                 NaN   
2                 NaN                NaN              NaN                 NaN   
3                 NaN                NaN              NaN                 NaN   
4                 1.0                NaN              NaN                 2.0   

   count_room_utility         date_created         date_updated  source  \
0                 NaN  2021-11-19 18:00:09  2021-11-19 17:44:45  baania   
1                 NaN  2022-01-27 14:30:04  2022-01-27 14:22:55  baania   
2                 NaN  2022-01-14 15:00:10  2022-01-14 14:40:04  baania   
3                 NaN  2020-12-04 10:40:59  2020-10-14 01:06:48  baania   
4                 NaN  2021-11-12 13:30:08  2021-02-03 12:12:54  baania   

                                        url_unittype  
0  https://baania.com/th/project/เฟ-ลิค-คอนโด-ลาด...  
1  https://baania.com/th/project/เซน-บาย-ศิวาลัย-...  
2  https://baania.com/th/project/วันพลัส-นายน์ทีน...  
3  https://baania.com/th/project/บ้านสวนชมน้ำ-776...  
4  https://baania.com/th/project/เบล็สทาวน์-ศรีนค...  

[5 rows x 27 columns]

# Data Clening

## Dealing with missing data

### Check missing data

In [6]:
project_df.describe()

row_number      latitude     longitude  subdistrict_id   district_id  \
count  23429.000000  23434.000000  23429.000000    23415.000000  23424.000000   
mean   11715.000000     16.399159    100.600086     4216.201751   4001.917435   
std     6763.514064    176.896370      0.947247     2476.968321   1662.485878   
min        1.000000      6.839354     97.985808     1007.000000   1006.000000   
25%     5858.000000     13.347151    100.400282     3414.000000   3411.000000   
50%    11715.000000     13.753725    100.601281     3802.000000   3800.000000   
75%    17572.000000     13.946742    100.912703     4527.000000   4000.000000   
max    23429.000000  12160.000000    105.222114    20688.000000  20003.000000   

        province_id        zipcode    count_unit  count_unittype  \
count  23425.000000   23403.000000  21547.000000    18603.000000   
mean    3942.592231   28209.221211    199.043765        2.325861   
std     1663.662562   24724.177495    320.029641        2.138378   
min        1.000000       1.000000      1.000000        1.000000   
25%     3372.000000   10400.000000     49.000000        1.000000   
50%     3781.000000   20000.000000    108.000000        2.000000   
75%     3781.000000   40000.000000    240.000000        3.000000   
max    10508.000000  390000.000000  10074.000000       96.000000   

       facility_clubhouse  facility_fitness  facility_meeting  facility_park  \
count         6820.000000       8895.000000       2890.000000   11675.000000   
mean             0.724340          0.848679          0.224913       0.894304   
std              0.446879          0.358382          0.417598       0.307461   
min              0.000000          0.000000          0.000000       0.000000   
25%              0.000000          1.000000          0.000000       1.000000   
50%              1.000000          1.000000          0.000000       1.000000   
75%              1.000000          1.000000          0.000000       1.000000   
max              1.000000          1.000000          1.000000       1.000000   

       facility_playground  facility_pool  facility_security  
count          6277.000000    9718.000000       16123.000000  
mean              0.697308       0.874048           0.959809  
std               0.459460       0.331812           0.196413  
min               0.000000       0.000000           0.000000  
25%               0.000000       1.000000           1.000000  
50%               1.000000       1.000000           1.000000  
75%               1.000000       1.000000           1.000000  
max               1.000000       1.000000           1.000000

วิธีเช็ควิธีที่ 1 ใช้ describe : จะเห็นว่า ค่า count จะสามารถช่วยเช็ค missing ได้ เนื่องจาก count จะนับเฉพาะแถวที่มีค่า เช่น
- อันที่มากที่สุด latatude = 23434
- แสดงว่า row number = 23429 จะมี missing 5 ตัว

In [7]:
project_df.isnull()

row_number  project_id  name_en  name_th  propertytype_id  \
0           False       False    False    False            False   
1           False       False    False    False            False   
2           False       False    False    False            False   
3           False       False    False    False            False   
4           False       False    False    False            False   
...           ...         ...      ...      ...              ...   
23429       False       False    False    False            False   
23430       False       False    False    False            False   
23431       False       False    False    False            False   
23432       False       False    False    False            False   
23433       False       False    False    False            False   

       propertytype_name_en  propertytype_name_th  price_min  developer_id  \
0                     False                 False      False         False   
1                     False                 False      False         False   
2                     False                 False      False         False   
3                     False                 False      False         False   
4                     False                 False      False         False   
...                     ...                   ...        ...           ...   
23429                 False                 False      False         False   
23430                 False                 False      False         False   
23431                 False                 False      False         False   
23432                 False                 False      False         False   
23433                 False                 False      False         False   

       developer_name_en  ...  facility_meeting  facility_park  \
0                  False  ...              True           True   
1                  False  ...              True          False   
2                   True  ...              True           True   
3                   True  ...              True           True   
4                  False  ...              True          False   
...                  ...  ...               ...            ...   
23429              False  ...              True          False   
23430              False  ...              True           True   
23431              False  ...              True          False   
23432              False  ...              True          False   
23433               True  ...              True           True   

       facility_playground  facility_pool  facility_security  date_created  \
0                     True          False              False         False   
1                     True          False              False         False   
2                     True           True               True         False   
3                     True           True               True         False   
4                     True          False              False         False   
...                    ...            ...                ...           ...   
23429                False           True              False         False   
23430                 True           True              False         False   
23431                False           True              False         False   
23432                False          False              False         False   
23433                 True           True               True         False   

       date_finish  date_updated  source  url_project  
0             True         False   False        False  
1            False         False   False        False  
2             True         False   False        False  
3            False         False   False        False  
4            False         False   False        False  
...            ...           ...     ...          ...  
23429        False         False   False        False  
23430        False         False   False        False  
23431        False 

In [8]:
project_df.isnull().any()

row_number                 True
project_id                False
name_en                   False
name_th                   False
propertytype_id           False
propertytype_name_en      False
propertytype_name_th      False
price_min                  True
developer_id              False
developer_name_en          True
developer_name_th          True
latitude                  False
longitude                  True
neighborhood_id            True
neighborhood_name_en       True
neighborhood_name_th       True
subdistrict_id             True
subdistrict_name_en        True
subdistrict_name_th        True
district_id                True
district_name_en           True
district_name_th           True
province_id                True
province_name_en           True
province_name_th           True
zipcode                    True
count_elevator             True
count_elevator_service     True
count_floor                True
count_parking              True
count_tower                True
count_un

In [9]:
project_df.isnull().any().any()

True

วิธีเช็ควิธีที่ 2 ใช้ isnull เพื่อดูค่าความจริง T (มี missing) / F (ไม่มี missing) แต่จะดูยาก

มักใช้ร่วมกับคำสั่ง any ซึ่งเป็นคำสั่งที่จะเอาค่าความจริงในแต่ละคอลัมน์มา or กัน

ถ้าใช้ any 2 รอบ จะเป็นการ or ทั้งตาราง เพื่อดูว่าข้อมูลที่ได้มามี missing หรือไม่

In [10]:
project_df.isnull().all()         # .all() เป็นการเอาค่าความจริงมา and กัน (ไม่ได้ใช้ในการหา missing แต่จะมีประโยชน์ในเรื่องอื่น)

row_number                False
project_id                False
name_en                   False
name_th                   False
propertytype_id           False
propertytype_name_en      False
propertytype_name_th      False
price_min                 False
developer_id              False
developer_name_en         False
developer_name_th         False
latitude                  False
longitude                 False
neighborhood_id           False
neighborhood_name_en      False
neighborhood_name_th      False
subdistrict_id            False
subdistrict_name_en       False
subdistrict_name_th       False
district_id               False
district_name_en          False
district_name_th          False
province_id               False
province_name_en          False
province_name_th          False
zipcode                   False
count_elevator            False
count_elevator_service    False
count_floor               False
count_parking             False
count_tower               False
count_un

### Delete records with missing data


```
dropna()
```

ข้อมูล missing: nul, Nan (not a number), Na (not authorised/ not available)

In [11]:
project_df.shape    # หาขนาดของ data

(23434, 45)

In [12]:
unit_df.shape

(42625, 27)

In [13]:
project_df.dropna()     # แถวไหนที่มี missing ให้ตัดทิ้ง

Empty DataFrame
Columns: [row_number, project_id, name_en, name_th, propertytype_id, propertytype_name_en, propertytype_name_th, price_min, developer_id, developer_name_en, developer_name_th, latitude, longitude, neighborhood_id, neighborhood_name_en, neighborhood_name_th, subdistrict_id, subdistrict_name_en, subdistrict_name_th, district_id, district_name_en, district_name_th, province_id, province_name_en, province_name_th, zipcode, count_elevator, count_elevator_service, count_floor, count_parking, count_tower, count_unit, count_unittype, facility_clubhouse, facility_fitness, facility_meeting, facility_park, facility_playground, facility_pool, facility_security, date_created, date_finish, date_updated, source, url_project]
Index: []

[0 rows x 45 columns]

In [14]:
unit_df.dropna()

Empty DataFrame
Columns: [row_number, unittype_id, project_id, name_th, propertytype_id, propertytype_name_en, propertytype_name_th, area_total_min_wa, area_usable_min, price_min, count_airconditioner, count_elevator, count_floor_internal, count_parking, count_room_bath, count_room_bed, count_room_dinning, count_room_guest, count_room_kitchen, count_room_living, count_room_maid, count_room_storage, count_room_utility, date_created, date_updated, source, url_unittype]
Index: []

[0 rows x 27 columns]

สามารถเข้าไปเช็คข้อมูลได้โดยพิมพ์เข้า google: dropna pandas --
จะเห็นว่า dropna เป็นฟังก์ชันที่มี input 5 ตัว (มี default ทั้งหมด)

DataFrame.dropna(axis=0, how='any', thresh=None, subset=None, inplace=False)
- axis: 0 = ดูตามแถว / 1 = ดูตามคอลัมน์
- how: any = ถ้ามี missing สักอัน จะตัด axis นั้นออกไป / all = ต้องเป็น missing ทั้ง axis ถึงจะตัดออก
- thresh (threshold): กำหนดจำนวน non-NA ถ้ามีน้อยกว่าที่กำหนดจะตัดแถวนั้นออกไป
- subset: พิจารณา missing เฉพาะคอลัมน์ที่กำหนด (ถ้ามีจะตัดแถวนั้นทิ้งไป) - เป็นการตัดโดยทำให้คอลัมน์ที่จะได้ใช้งานต่อมีค่าเสมอ
- inplace: true = อัปเดตตัวแปรเดิมให้เป็นตัวใหม่ที่ทำ (มีค่าเท่ากับการเอาตัวแปรใหม่มารับ) -- ปกติไม่ค่อยใช้ เพราะจะทำให้ไล่ยาก ด้วยความไม่รู้ว่าตัวแปรปัจจุบันถูกเปลี่ยนไปเป็นอะไรตั้งแต่เมื่อไหร่ มักจะเอาตัวแปรอื่นมารับมากกว่า

In [15]:
project_df.dropna(axis=1)     # หลักไหนมี missing ให้ตัดทิ้ง

project_id                                       name_en  \
0      5ee08d64dda791001299b2a6            Winning Residence Sukhumvit Praksa   
1                 project_11326                              Gift Condominium   
2      62739c4089233e001362b001              Sanrak 9 Sukprayoon Nong Krathum   
3                  project_2010                               Baan Sansabai 1   
4                 project_16838                    Baanpruksa Ratchapruek-345   
...                         ...                                           ...   
23429             project_12317  Baan Klangmeung Urbanion Ladphrao-Chokchai 4   
23430             project_12057                           @Home Kaset-Nawamin   
23431              project_4461                    I-Design Wongwaen-Ramintra   
23432               project_144                 Koolpunt Ville 15 Park Avenue   
23433             project_16300                          Thawisuk Garden Home   

                                             name_th propertytype_id  \
0                 วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1   
1                                กิ๊ฟท์ คอนโดมิเนียม               2   
2                     แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3   
3                                      บ้านแสนสบาย 1               1   
4                             บ้านพฤกษา ราชพฤกษ์-345               3   
...                                              ...             ...   
23429  บ้านกลางเมือง เออร์บาเนี่ยน ลาดพร้าว-โชคชัย 4               3   
23430                          แอทโฮม เกษตร-นวมินทร์               3   
23431                     ไอ-ดีไซน์ วงแหวน-รามอินทรา               4   
23432                  กุลพันธ์วิลล์ 15 พาร์ค อเวนิว               1   
23433                             ทวีสุข การ์เด้นโฮม               1   

      propertytype_name_en propertytype_name_th     developer_id   latitude  
0           Detached House                 บ้าน  developer_10678  13.570530  
1                    Condo                คอนโด   developer_7319  13.033814  
2                 Townhome             ทาวน์โฮม  developer_10510  13.417402  
3           Detached House                 บ้าน  developer_10510  15.038042  
4                 Townhome             ทาวน์โฮม   developer_1026  13.965246  
...                    ...                  ...              ...        ...  
23429             Townhome             ทาวน์โฮม   developer_1760  13.814468  
23430             Townhome             ทาวน์โฮม   developer_1793  13.834361  
23431           Commercial         อาคารพาณิชย์   developer_2217  13.859872  
23432       Detached House                 บ้าน   developer_1004  18.808447  
23433       Detached House                 บ้าน  developer_10510  12.684945  

[23434 rows x 8 columns]

In [16]:
project_df.dropna(axis=1).isnull().any().any()

False

In [17]:
print(project_df.isnull().any().any())
print(project_df.shape)

True
(23434, 45)


คำสั่งอะไรที่เมื่อรันแล้วได้ผลแสดงออกมาบนหน้าจอ จะเป็นการจบคำสั่งแล้ว ไม่สามารถเอาไปใช้งานต่อได้

หากต้องการใช้งานต่อ จะต้องเอาตัวแปรมารับ

In [19]:
project_df2 = project_df.dropna(axis=1)
project_df2.head()

project_id                             name_en  \
0  5ee08d64dda791001299b2a6  Winning Residence Sukhumvit Praksa   
1             project_11326                    Gift Condominium   
2  62739c4089233e001362b001    Sanrak 9 Sukprayoon Nong Krathum   
3              project_2010                     Baan Sansabai 1   
4             project_16838          Baanpruksa Ratchapruek-345   

                              name_th propertytype_id propertytype_name_en  \
0  วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1       Detached House   
1                 กิ๊ฟท์ คอนโดมิเนียม               2                Condo   
2      แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3             Townhome   
3                       บ้านแสนสบาย 1               1       Detached House   
4              บ้านพฤกษา ราชพฤกษ์-345               3             Townhome   

  propertytype_name_th     developer_id   latitude  
0                 บ้าน  developer_10678  13.570530  
1                คอนโด   developer_7319  13.033814  
2             ทาวน์โฮม  developer_10510  13.417402  
3                 บ้าน  developer_10510  15.038042  
4             ทาวน์โฮม   developer_1026  13.965246

In [22]:
project_df.iloc[:5,-7:]       # iloc = คำสั่งตัดแถว/คอลัมน์ จากตารางใหญ่ (มองตารางเป็นเมทริกซ์ และทำ table slicing)

facility_pool  facility_security         date_created          date_finish  \
0            1.0                1.0  2020-06-10 14:36:04                  NaN   
1            1.0                1.0  2017-10-30 15:38:48  2013-01-01 07:00:00   
2            NaN                NaN  2022-05-05 16:43:28                  NaN   
3            NaN                NaN  2016-09-05 15:32:15  1970-01-02 07:00:00   
4            1.0                1.0  2019-09-05 11:49:37  1970-01-02 07:00:00   

          date_updated  source  \
0  2020-12-04 10:32:58  baania   
1  2020-12-04 10:49:00  baania   
2  2022-05-05 17:30:03  baania   
3  2020-12-04 10:45:31  baania   
4  2021-03-17 12:00:04  baania   

                                         url_project  
0  https://baania.com/th/project/วินนิ่ง-เรสซิเดน...  
1  https://baania.com/th/project/กิ๊ฟท์-คอนโดมิเน...  
2  https://baania.com/th/project/แสนรัก-9-ศุขประย...  
3  https://baania.com/th/project/บ้านแสนสบาย-1-pr...  
4  https://baania.com/th/project/บ้านพฤกษา-ราชพฤก...

In [23]:
project_df.iloc[:5,-7:].dropna(thresh=5)    #ตัดแถวที่มีค่าที่ไม่ใช่ missing น้อยกว่า 5 ตัว

facility_pool  facility_security         date_created          date_finish  \
0            1.0                1.0  2020-06-10 14:36:04                  NaN   
1            1.0                1.0  2017-10-30 15:38:48  2013-01-01 07:00:00   
3            NaN                NaN  2016-09-05 15:32:15  1970-01-02 07:00:00   
4            1.0                1.0  2019-09-05 11:49:37  1970-01-02 07:00:00   

          date_updated  source  \
0  2020-12-04 10:32:58  baania   
1  2020-12-04 10:49:00  baania   
3  2020-12-04 10:45:31  baania   
4  2021-03-17 12:00:04  baania   

                                         url_project  
0  https://baania.com/th/project/วินนิ่ง-เรสซิเดน...  
1  https://baania.com/th/project/กิ๊ฟท์-คอนโดมิเน...  
3  https://baania.com/th/project/บ้านแสนสบาย-1-pr...  
4  https://baania.com/th/project/บ้านพฤกษา-ราชพฤก...

In [31]:
project_df.iloc[:5,-7:].dropna(subset=['facility_pool'])      #พิจารณา missing ในคอลัมน์ facility_pool ถ้ามีก็ตัดทิ้ง

facility_pool  facility_security         date_created          date_finish  \
0            1.0                1.0  2020-06-10 14:36:04                  NaN   
1            1.0                1.0  2017-10-30 15:38:48  2013-01-01 07:00:00   
4            1.0                1.0  2019-09-05 11:49:37  1970-01-02 07:00:00   

          date_updated  source  \
0  2020-12-04 10:32:58  baania   
1  2020-12-04 10:49:00  baania   
4  2021-03-17 12:00:04  baania   

                                         url_project  
0  https://baania.com/th/project/วินนิ่ง-เรสซิเดน...  
1  https://baania.com/th/project/กิ๊ฟท์-คอนโดมิเน...  
4  https://baania.com/th/project/บ้านพฤกษา-ราชพฤก...

In [32]:
project_df.dropna(subset=['propertytype_name_en', 'price_min'])

row_number                project_id  \
0          7880.0  5ee08d64dda791001299b2a6   
1         16689.0             project_11326   
2           290.0  62739c4089233e001362b001   
3         20719.0              project_2010   
4           199.0             project_16838   
...           ...                       ...   
23429     12883.0             project_12317   
23430     21181.0             project_12057   
23431     13993.0              project_4461   
23432      5889.0               project_144   
23433      5529.0             project_16300   

                                            name_en  \
0                Winning Residence Sukhumvit Praksa   
1                                  Gift Condominium   
2                  Sanrak 9 Sukprayoon Nong Krathum   
3                                   Baan Sansabai 1   
4                        Baanpruksa Ratchapruek-345   
...                                             ...   
23429  Baan Klangmeung Urbanion Ladphrao-Chokchai 4   
23430                           @Home Kaset-Nawamin   
23431                    I-Design Wongwaen-Ramintra   
23432                 Koolpunt Ville 15 Park Avenue   
23433                          Thawisuk Garden Home   

                                             name_th propertytype_id  \
0                 วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1   
1                                กิ๊ฟท์ คอนโดมิเนียม               2   
2                     แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3   
3                                      บ้านแสนสบาย 1               1   
4                             บ้านพฤกษา ราชพฤกษ์-345               3   
...                                              ...             ...   
23429  บ้านกลางเมือง เออร์บาเนี่ยน ลาดพร้าว-โชคชัย 4               3   
23430                          แอทโฮม เกษตร-นวมินทร์               3   
23431                     ไอ-ดีไซน์ วงแหวน-รามอินทรา               4   
23432                  กุลพันธ์วิลล์ 15 พาร์ค อเวนิว               1   
23433                             ทวีสุข การ์เด้นโฮม               1   

      propertytype_name_en propertytype_name_th  price_min     developer_id  \
0           Detached House                 บ้าน  3790000.0  developer_10678   
1                    Condo                คอนโด  1100000.0   developer_7319   
2                 Townhome             ทาวน์โฮม        0.0  developer_10510   
3           Detached House                 บ้าน        0.0  developer_10510   
4                 Townhome             ทาวน์โฮม  1930000.0   developer_1026   
...                    ...                  ...        ...              ...   
23429             Townhome             ทาวน์โฮม  5700000.0   developer_1760   
23430             Townhome             ทาวน์โฮม  3590000.0   developer_1793   
23431           Commercial         อาคารพาณิชย์  1870000.0   developer_2217   
23432       Detached House                 บ้าน  4000000.0   developer_1004   
23433       Detached House                 บ้าน  2800000.0  developer_10510   

                   developer_name_en  ... facility_meeting  facility_park  \
0       Winning Asset Plus Co., Ltd.  ...              NaN            NaN   
1           Gift Asset Land Co.,Ltd.  ...              NaN            1.0   
2                                NaN  ...              NaN            NaN   
3                                NaN  ...              NaN            NaN   
4       Preuksa Real Estate Co.,Ltd.  ...              NaN            1.0   
...                              ...  ...              ...            ...   
23429         AP (Thailand) Co,. Ltd  ...              NaN            1.0   
23430       Areeya Property Co.,Ltd.  ...              NaN            NaN   
23431  Wangthong Group Pub Co., Ltd.  ...              NaN            1.0   
23432     Homeland Service Co., Ltd.  ...              NaN            1.0   
23433                            NaN  ...              NaN            NaN   

       facility_playground facility_pool f

In [33]:
project_df.dropna(subset=['propertytype_name_en', 'price_min']).isnull().any()      #จะเห็นว่าในสองคอลัมน์ที่กำหนด จะให้ค่าเป็น false

row_number                 True
project_id                False
name_en                   False
name_th                   False
propertytype_id           False
propertytype_name_en      False
propertytype_name_th      False
price_min                 False
developer_id              False
developer_name_en          True
developer_name_th          True
latitude                  False
longitude                  True
neighborhood_id            True
neighborhood_name_en       True
neighborhood_name_th       True
subdistrict_id             True
subdistrict_name_en        True
subdistrict_name_th        True
district_id                True
district_name_en           True
district_name_th           True
province_id                True
province_name_en           True
province_name_th           True
zipcode                    True
count_elevator             True
count_elevator_service     True
count_floor                True
count_parking              True
count_tower                True
count_un

In [36]:
# project_df.dropna(subset=['propertytype_name_en', 'price_min'],inplace=True)

#### Quiz3

1. กำจัดค่า missing โดยการตัดคอลัมน์ให้น้อยที่สุด โดยที่จำนวน records ที่ไม่มี missing ยังเลยเหลือมากกว่า 80%
2. สร้าง data ที่ค่าในคอลัมน์ [price_min, neighborhood_name_en, developer_name_en] มี missing ไม่เกิน 1 ค่า

In [56]:
project_df = pd.read_csv('/content/drive/MyDrive/DPDM2022/Copy of opendata_project.csv')

In [58]:
# ข้อ 1
project_df.dropna(axis=1, thresh=0.8*project_df.shape[0]) 

row_number                project_id  \
0          7880.0  5ee08d64dda791001299b2a6   
1         16689.0             project_11326   
2           290.0  62739c4089233e001362b001   
3         20719.0              project_2010   
4           199.0             project_16838   
...           ...                       ...   
23429     12883.0             project_12317   
23430     21181.0             project_12057   
23431     13993.0              project_4461   
23432      5889.0               project_144   
23433      5529.0             project_16300   

                                            name_en  \
0                Winning Residence Sukhumvit Praksa   
1                                  Gift Condominium   
2                  Sanrak 9 Sukprayoon Nong Krathum   
3                                   Baan Sansabai 1   
4                        Baanpruksa Ratchapruek-345   
...                                             ...   
23429  Baan Klangmeung Urbanion Ladphrao-Chokchai 4   
23430                           @Home Kaset-Nawamin   
23431                    I-Design Wongwaen-Ramintra   
23432                 Koolpunt Ville 15 Park Avenue   
23433                          Thawisuk Garden Home   

                                             name_th propertytype_id  \
0                 วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1   
1                                กิ๊ฟท์ คอนโดมิเนียม               2   
2                     แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3   
3                                      บ้านแสนสบาย 1               1   
4                             บ้านพฤกษา ราชพฤกษ์-345               3   
...                                              ...             ...   
23429  บ้านกลางเมือง เออร์บาเนี่ยน ลาดพร้าว-โชคชัย 4               3   
23430                          แอทโฮม เกษตร-นวมินทร์               3   
23431                     ไอ-ดีไซน์ วงแหวน-รามอินทรา               4   
23432                  กุลพันธ์วิลล์ 15 พาร์ค อเวนิว               1   
23433                             ทวีสุข การ์เด้นโฮม               1   

      propertytype_name_en propertytype_name_th  price_min     developer_id  \
0           Detached House                 บ้าน  3790000.0  developer_10678   
1                    Condo                คอนโด  1100000.0   developer_7319   
2                 Townhome             ทาวน์โฮม        0.0  developer_10510   
3           Detached House                 บ้าน        0.0  developer_10510   
4                 Townhome             ทาวน์โฮม  1930000.0   developer_1026   
...                    ...                  ...        ...              ...   
23429             Townhome             ทาวน์โฮม  5700000.0   developer_1760   
23430             Townhome             ทาวน์โฮม  3590000.0   developer_1793   
23431           Commercial         อาคารพาณิชย์  1870000.0   developer_2217   
23432       Detached House                 บ้าน  4000000.0   developer_1004   
23433       Detached House                 บ้าน  2800000.0  developer_10510   

        latitude  ...  province_id   province_name_en province_name_th  \
0      13.570530  ...       3498.0       Samut Prakan      สมุทรปราการ   
1      13.033814  ...       4616.0          Chon Buri           ชลบุรี   
2      13.417402  ...       4616.0          Chon Buri           ชลบุรี   
3      15.038042  ...       2857.0  Nakhon Ratchasima       นครราชสีมา   
4      13.965246  ...       3599.0       Pathum Thani         ปทุมธานี   
...          ...  ...          ...                ...              ...   
23429  13.814468  ...       3781.0            Bangkok    กรุงเทพมหานคร   
23430  13.834361  ...       3781.0            Bangkok    กรุงเทพมหานคร   
23431  13.859872  ...       3781.0            Bangkok    กรุงเทพมหานคร   
23432  18.808447  ...       1005.0         Chiang Mai        เชียงใหม่   
23433  12.684945  ...       3431.0             Rayong            ระยอง   

       zipcode  count_unit         date_created          date_finish  \
0     

In [59]:
# check - เมื่อ dropna() จะต้องมีจำนวนแถวเกิน 80% = 18747.2
project_df.dropna(axis=1, thresh=0.8*project_df.shape[0]).dropna()

row_number     project_id  \
1         16689.0  project_11326   
4           199.0  project_16838   
5          9434.0  project_15884   
7          9022.0   project_1488   
8          7266.0   project_6637   
...           ...            ...   
23429     12883.0  project_12317   
23430     21181.0  project_12057   
23431     13993.0   project_4461   
23432      5889.0    project_144   
23433      5529.0  project_16300   

                                            name_en  \
1                                  Gift Condominium   
4                        Baanpruksa Ratchapruek-345   
5                                           H2 Wood   
7                        Homeland 3 Paradise Garden   
8                      Lumpini Place Rama 4-Sathorn   
...                                             ...   
23429  Baan Klangmeung Urbanion Ladphrao-Chokchai 4   
23430                           @Home Kaset-Nawamin   
23431                    I-Design Wongwaen-Ramintra   
23432                 Koolpunt Ville 15 Park Avenue   
23433                          Thawisuk Garden Home   

                                             name_th propertytype_id  \
1                                กิ๊ฟท์ คอนโดมิเนียม               2   
4                             บ้านพฤกษา ราชพฤกษ์-345               3   
5                                          เอชทู วูด               2   
7                       โฮมแลนด์ 3 พาราไดซ์ การ์เด้น               1   
8                           ลุมพินีเพลสพระราม 4-สาทร               2   
...                                              ...             ...   
23429  บ้านกลางเมือง เออร์บาเนี่ยน ลาดพร้าว-โชคชัย 4               3   
23430                          แอทโฮม เกษตร-นวมินทร์               3   
23431                     ไอ-ดีไซน์ วงแหวน-รามอินทรา               4   
23432                  กุลพันธ์วิลล์ 15 พาร์ค อเวนิว               1   
23433                             ทวีสุข การ์เด้นโฮม               1   

      propertytype_name_en propertytype_name_th  price_min     developer_id  \
1                    Condo                คอนโด  1100000.0   developer_7319   
4                 Townhome             ทาวน์โฮม  1930000.0   developer_1026   
5                    Condo                คอนโด  1480000.0   developer_1789   
7           Detached House                 บ้าน  2600000.0   developer_1336   
8                    Condo                คอนโด  1990000.0   developer_1745   
...                    ...                  ...        ...              ...   
23429             Townhome             ทาวน์โฮม  5700000.0   developer_1760   
23430             Townhome             ทาวน์โฮม  3590000.0   developer_1793   
23431           Commercial         อาคารพาณิชย์  1870000.0   developer_2217   
23432       Detached House                 บ้าน  4000000.0   developer_1004   
23433       Detached House                 บ้าน  2800000.0  developer_10510   

        latitude  ...  province_id   province_name_en province_name_th  \
1      13.033814  ...       4616.0          Chon Buri           ชลบุรี   
4      13.965246  ...       3599.0       Pathum Thani         ปทุมธานี   
5      13.865311  ...       3781.0            Bangkok    กรุงเทพมหานคร   
7      14.973075  ...       2857.0  Nakhon Ratchasima       นครราชสีมา   
8      13.718633  ...       3781.0            Bangkok    กรุงเทพมหานคร   
...          ...  ...          ...                ...              ...   
23429  13.814468  ...       3781.0            Bangkok    กรุงเทพมหานคร   
23430  13.834361  ...       3781.0            Bangkok    กรุงเทพมหานคร   
23431  13.859872  ...       3781.0            Bangkok    กรุงเทพมหานคร   
23432  18.808447  ...       1005.0         Chiang Mai        เชียงใหม่   
23433  12.684945  ...       3431.0             Rayong            ระยอง   

       zipcode  count_unit         date_created          date_finish  \
1      20230.0        77.0  2017-10-30 15:38:48  2013-01-01 07:00:00   
4      12000.0       485.0  2019-09-05 11:49:37  1970-01-02 07:00:

In [57]:
# ข้อ 2
project_df.dropna(subset=['price_min', 'neighborhood_name_en', 'developer_name_en'], thresh=2) 

row_number                project_id  \
0          7880.0  5ee08d64dda791001299b2a6   
1         16689.0             project_11326   
2           290.0  62739c4089233e001362b001   
3         20719.0              project_2010   
4           199.0             project_16838   
...           ...                       ...   
23429     12883.0             project_12317   
23430     21181.0             project_12057   
23431     13993.0              project_4461   
23432      5889.0               project_144   
23433      5529.0             project_16300   

                                            name_en  \
0                Winning Residence Sukhumvit Praksa   
1                                  Gift Condominium   
2                  Sanrak 9 Sukprayoon Nong Krathum   
3                                   Baan Sansabai 1   
4                        Baanpruksa Ratchapruek-345   
...                                             ...   
23429  Baan Klangmeung Urbanion Ladphrao-Chokchai 4   
23430                           @Home Kaset-Nawamin   
23431                    I-Design Wongwaen-Ramintra   
23432                 Koolpunt Ville 15 Park Avenue   
23433                          Thawisuk Garden Home   

                                             name_th propertytype_id  \
0                 วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1   
1                                กิ๊ฟท์ คอนโดมิเนียม               2   
2                     แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3   
3                                      บ้านแสนสบาย 1               1   
4                             บ้านพฤกษา ราชพฤกษ์-345               3   
...                                              ...             ...   
23429  บ้านกลางเมือง เออร์บาเนี่ยน ลาดพร้าว-โชคชัย 4               3   
23430                          แอทโฮม เกษตร-นวมินทร์               3   
23431                     ไอ-ดีไซน์ วงแหวน-รามอินทรา               4   
23432                  กุลพันธ์วิลล์ 15 พาร์ค อเวนิว               1   
23433                             ทวีสุข การ์เด้นโฮม               1   

      propertytype_name_en propertytype_name_th  price_min     developer_id  \
0           Detached House                 บ้าน  3790000.0  developer_10678   
1                    Condo                คอนโด  1100000.0   developer_7319   
2                 Townhome             ทาวน์โฮม        0.0  developer_10510   
3           Detached House                 บ้าน        0.0  developer_10510   
4                 Townhome             ทาวน์โฮม  1930000.0   developer_1026   
...                    ...                  ...        ...              ...   
23429             Townhome             ทาวน์โฮม  5700000.0   developer_1760   
23430             Townhome             ทาวน์โฮม  3590000.0   developer_1793   
23431           Commercial         อาคารพาณิชย์  1870000.0   developer_2217   
23432       Detached House                 บ้าน  4000000.0   developer_1004   
23433       Detached House                 บ้าน  2800000.0  developer_10510   

                   developer_name_en  ... facility_meeting  facility_park  \
0       Winning Asset Plus Co., Ltd.  ...              NaN            NaN   
1           Gift Asset Land Co.,Ltd.  ...              NaN            1.0   
2                                NaN  ...              NaN            NaN   
3                                NaN  ...              NaN            NaN   
4       Preuksa Real Estate Co.,Ltd.  ...              NaN            1.0   
...                              ...  ...              ...            ...   
23429         AP (Thailand) Co,. Ltd  ...              NaN            1.0   
23430       Areeya Property Co.,Ltd.  ...              NaN            NaN   
23431  Wangthong Group Pub Co., Ltd.  ...              NaN            1.0   
23432     Homeland Service Co., Ltd.  ...              NaN            1.0   
23433                            NaN  ...              NaN            NaN   

       facility_playground facility_pool f

***